### PR TITLE
DRAFT: Load LWOLoader.js from the github.com/threejs/lwoloader repo

### PR DESCRIPTION
**Pull request by looeee**
_Note: this PR is not intended to be merged, for discussion only_

Recently myself and @sciecode moved the development of the `LWOLoader` out of this repo and into it's own repo under the three.js org: [github.com/threejs/lwoloader](https://github.com/threejs/lwoloader). 

This has given us a whole range of benefits, such as a proper testing environment with lots of [LWO test models](https://raw.githack.com/threejs/lwoloader/master/index.html). We were also able to adopt a modern, modular approach to developing the loader rather than having to keep everything in one file. 

But, we now have two versions of the loader. One on this repo, and one on the remote repo. Not only that, but with the modularize script here, we now are developing the loader using modules, bundling them into a single ES5 compatible script file to put in the `examples/js` directory... and _then_ running `modularize.js` on that to get back an ES6 module! Pretty crazy 😅 

This is a quick demo of how we can get around that and load [LWOLoader.js](https://github.com/threejs/lwoloader/blob/master/build/js/LWOLoader.js) script directly from the remote repo, without making any major changes to the example file (just a single line changed). 

Note: I'm using raw.githack for this quick demo. A better approach would be to set up github pages for the [github.com/threejs](https://github.com/threejs) org and serve the files from there. 

### [Live](https://raw.githack.com/looeee/three.js/lwoloader-remote-example-test/examples/webgl_loader_lwo.html)

### CODE REVIEW

**Comment by mrdoob** [comment]
I'm not a big fan of this suggestion. It may not be that important anymore, but I always wanted the repo to be self-contained. You can download it and it's ready to go, you no longer need internet.

**Comment by looeee** [comment]
> I'm not a big fan of this suggestion. 

That's fine, I just opened this to illustrate one possible solution. A PR is worth a thousand words as they say.

>  I always wanted the repo to be self-contained. You can download it and it's ready to go

Yeah, I get that. It's certainly handy, especially for developers. For users it's less and less common to download an entire github repo though - in fact, it's already becoming frustrating for users as the download size grows (see #14457). 

A better approach, I think, would be to direct people to install the NPM module if they want to use three.js, and add an online code editor to the examples so that people can experiment with three.js without having to  download anything, or figure out how to run an HTTP server (see the [Babylon.js playground](https://playground.babylonjs.com/#BCU1XR#0)). 

> you no longer need internet.

Is that something we need to worry about? I mean, I live on a tiny island in Thailand at the moment and I still have a 200mbs connection. Even when the power goes the internet stays on.

**Comment by gkjohnson** [comment]
I really like it in a single repo, as well. Aside from those already mentioned the pragmatic issues associated with maintaining and updating what could be dozens of repos is a lot take on.

With all the code in a single repo it makes it much easier to keep example code in sync and make sure that breaking changes in core are properly updated in all the example files. Rather than publish a new breaking three.js build, update to it in the loader repo, fix the issues, then publish that package we can fix everything and publish it in one go and users can trust that all the files work in a new release. Maintaining the examples site would probably also get more complicated.

The ES6 modular-ization of code is also a bit of a road block because at the moment the example modules can reference the three.js module via relative file path but they won't be able to if the loader is in another package (without a build process). It's a (pretty massive) limitation of es6 modules at the moment but unfortunately that's how it is right now. If we want to enable easy, out of the box use of the example files I'm not sure how that would work.

If we wanted to afford more test-ability or maintainability and individual example files (tests, etc) then I would think something closer to a monorepo could be a better fit but there might still be some quirks to work out.

**Comment by donmccurdy** [comment]
If the primary authors/maintainers of an example module agree that it would be better off in a different repository, I think that's their prerogative, but IMO the module should no longer be referenced by three.js docs, examples, or tests in that case, except perhaps linking to the project page somewhere. There are plenty of modules that fit this description:

- https://github.com/Jam3/three-bmfont-text
- https://github.com/zadvorsky/three.bas
- https://github.com/chandlerprall/ThreeCSG
- https://github.com/donmccurdy/three-pathfinding

The benefits — e.g. full control of build and test processes — are most worthwhile to large/complex modules, I would think. Having each loader in a separate repository does not scale to all loaders, so I'm definitely hesitant about the specific suggestion here.  

The idea that more repositories would mean faster turnaround on PRs doesn't seem obvious to me – the otherwise solid PRs that remain open tend to be ones where no one is confident saying the change is good, or no one has time to review and improve it. If a module has enough active maintainers to run it themselves as a separate repository, without relying on the core committers, that is probably not a problem. On the other hand, it will be harder to confidently merge PRs to the main repository if it's harder to test the implications on example modules.

> We were also able to adopt a modern, modular approach to developing the loader rather than having to keep everything in one file.

At the end of the ES modularization journey, I'd like to think this will be possible within `examples/jsm` and the `examples/js` versions could be generated (outputting a single file) from that.

**Comment by looeee** [comment]
>  It's a (pretty massive) limitation of es6 modules at the moment but unfortunately that's how it is right now.

Yeah, I just realised that yesterday while converting some of the examples to modules 🤔 

> Having each loader in a separate repository does not scale to all loaders, so I'm definitely hesitant about the specific suggestion here.
> tl;dr – I'd lean toward as few repositories as possible, ideally just one.

I'm not keen on the idea of going the D3 route and having 100s of tiny, fragmented repos either. But there is the middle ground of having just a couple of medium sized ones:

* core
* loaders
* editor
* etc.

**Comment by Mugen87** [comment]
I suggest to close the PR and #16679. It seems the majority of the project members prefer a single repo.

**Comment by looeee** [comment]
> It seems the majority of the project members prefer a single repo.

The majority of people haven't said anything at all, which is not the same as dissent. I'm fine with closing this and #16679, but only if these threads don't get referenced as a way of silencing future discussion on this topic. 

Regarding the `LWOLoader`, I'll move it into this repo once we're treating the module versions of the loaders as the canonical version. We can keep the other repo for test models and reference the loader from here.
